### PR TITLE
docs: Strat the technical notes

### PR DESCRIPTION
**Pull request by TheoRudkiewicz**
Some technical notes to explain the internal design of GroMo.

### CODE REVIEW

File: docs/source/tech_notes.rst [hunk 1]
```diff
@@ -0,0 +1,59 @@
+GrowingModule (GroMo)
+=====================
+
+This is the documentation of the FoGro method described in [TMLR24] Verbockhaven, M., Rudkiewicz, T., Chevallier, S., and Charpiat, G. (2024). Growing tiny networks: Spotting expressivity bottlenecks and fixing them optimally. TMLR..
```
Comment (sylvchev) [inline docs/source/tech_notes.rst:4]:
```suggestion
This is the documentation of the growing method described in [TMLR24]_.
```

FoGro is not mentioned in the paper, better to put the reference at the end.

File: docs/source/tech_notes.rst [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+- :math:`\texttt{extended\_forward}(x) = (M1(x), \gamma  O1(x))`
+- :math:`\texttt{extended\_forward}(M1(x), \gamma  O1(x)) = M2(M1(x)) + \gamma^2  I2(O1(x))`
+
+Once your satisfied with the extension and the scaling factor you incorporate it in the forward pass by using the method `apply_change` in `M2` with the option `apply_previous` set to `True`.
+
+=====================
+Glossary
+=====================
+
+- machine batch / statistical batch: often in ML we use batch of data to estimate quantities and to process data together to make computation faster. For example in stochastic gradient descent we estimate the gradient with on batch of example that is computed at the same time. In our case we can compute multiple examples at the same time and do it multiple times to compute a statistics. We refer to the first one to machine batch and to the second as statistical batch.
```
Comment (sylvchev) [inline docs/source/tech_notes.rst:59]:
```suggestion
- machine batch / statistical batch: often in ML we use batch of data to estimate quantities and to process data together to make computation faster. For example in stochastic gradient descent we estimate the gradient on a batch of examples that is computed at the same time. In our case we can compute multiple examples at the same time and do it multiple times to compute a statistics. We refer to the first one to machine batch and to the second as statistical batch.

=========
Bibliography
=========

.. [TMLR24] Verbockhaven, M., Rudkiewicz, T., Chevallier, S., and Charpiat, G.
        (2024). Growing tiny networks: Spotting expressivity bottlenecks and
        fixing them optimally. TMLR.

```

We need to use a standardize convention with Stella for naming the batch.

File: docs/source/tech_notes.rst [hunk 3]
```diff
@@ -0,0 +1,59 @@
+GrowingModule (GroMo)
+=====================
+
+This is the documentation of the FoGro method described in [TMLR24] Verbockhaven, M., Rudkiewicz, T., Chevallier, S., and Charpiat, G. (2024). Growing tiny networks: Spotting expressivity bottlenecks and fixing them optimally. TMLR..
+
+We propose to build growing networks using `GroMo` inherited from `torch.nn.Module`. We have two type of modules, the standard `GrowingModule` and the `AdditionGrowingModule`. The first one is a connections that contains the parameters of the network. The second one aims to connect multiple `GrowingModule` together.
```
Comment (sylvchev) [inline docs/source/tech_notes.rst:6]:
```suggestion
We propose to build growing networks using `GroMo` inherited from `torch.nn.Module`. We have two types of modules, the standard `GrowingModule` and the `AdditionGrowingModule`. The first one is a connection that contains the parameters of the network. The second one aims to connect multiple `GrowingModule` together.
```

**Review by sylvchev — COMMENTED** [review]
math are not rendered correctly.

**Comment by TheoRudkiewicz** [comment]
Thank you for the typo correction.
Indeed math is not rendering on the github compiled version. However it is working with the .rst interpreter of my vscode and seems to be a valid syntax for latex in .rst. I need to check what is the correct syntax or why it is not working on the github side. If anyone has any clue I am interested.

**Comment by sylvchev** [comment]
@TheoRudkiewicz You will need to update your branch, I'll try to solve the issue of math rendering

**Comment by sylvchev** [comment]
math are correctly rendered now 🎉

File: docs/source/tech_notes.rst [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+Sequential case
+----------------
+
+To grow a layer of neurons we need new incoming connections and new outgoing connections. If we look a at section with two modules `M1` and `M2` with a layer of neurons in between we can grow the layer of neurons by adding new outgoing connections in `M1` (noted `O1`) and new incoming connections in `M2` (noted `I2`). To compute the new weights with the TINY/FoGro method we can use the method `compute_optimal_added_parameters` in `M2` that will also update the module `M1` if the option `update_previous` is set to `True`.
+
+You can then use the `extended_forward` method to compute the forward pass with the new weights. More precisely, the `extended_forward` with integrate the `scaling_factor` (noted :math:`\gamma`) to the forward pass:
+
+- :math:`\texttt{extended\_forward}(x) = (M1(x), \gamma  O1(x))`
+- :math:`\texttt{extended\_forward}(M1(x), \gamma  O1(x)) = M2(M1(x)) + \gamma^2  I2(O1(x))`
```
Comment (sylvchev) [inline docs/source/tech_notes.rst:51]: extended_forward is incorrectly rendered with the \

Comment (TheoRudkiewicz) [reply docs/source/tech_notes.rst]: Yes, I fixed it !

**Comment by TheoRudkiewicz** [comment]
> @TheoRudkiewicz You will need to update your branch, I'll try to solve the issue of math rendering
> 
> You mark this PR as a draft, do you plan to add other things in this PR or could we merge when we solve the math rendering issue?

Thank you for correction of  the math rendering.
I put the PR in draft the time to fix it.
It is now ready to merge.

**Comment by sylvchev** [comment]
It seems notification are not pushed on my phone. Could you ping me on MM when you need review?

**Review by sylvchev — APPROVED** [review]

**Comment by sylvchev** [comment]
Thanks @TheoRudkiewicz 🙏

**Pull request merged**